# this is the snakemake pipeline for analysis the metagenomic data

configfile: "config/config.yml"

BARCODES = ["barcode01", "barcode02", "barcode03", "barcode04", "barcode05", "barcode06"]

rule all:
    input:
        "results/barcoding_summary.txt",
        expand("results/debarcoded/duplex/{barcode}.fastq", barcode = BARCODES)


rule dorado_basecall:
    output:
        bam = "results/dorado/output.bam"
    threads: 4
    params:
        input_dir = config["raw_data_dir"],
        model = config["dorado_model"],
        dorado = config["dorado_path"]
    resources:
        mem_mb = 8000,
        partition = "gpu",
        slurm = "gpus=1"
    shell:
        """
        {params.dorado} basecaller -r {params.model} {params.input_dir} > {output.bam}
        """

rule sort_bam:
    input:
        "results/dorado/output.bam"
    output:
        "results/dorado/sorted/output.sorted.bam"
    conda:
        "envs/readtools.yml"
    threads: 16
    resources:
        mem_mb = 16000
    shell:
        """
        samtools sort -@ {threads} -o {output} {input}
        """

rule debarcode:
    input:
        "results/dorado/sorted/output.sorted.bam"
    output:
        "results/debarcoded/barcoding_summary.txt",
        expand("results/debarcoded/barcode{barcode}/fastq_runid_unknown_0.fastq", barcode = config["barcodes"])
    threads: 4
    resources:
        mem_mb = 450000,
        partition = "gpu",
        slurm = "gpus=1"
    shell:
        """
        $APPS/ont-guppy/bin/guppy_barcoder -i results/dorado/sorted/ --save_path results/debarcoded/ --barcode_kits SQK-NBD114-24 -t 4 -q 0 --enable_trim_barcodes
        """

rule extract_readID:
    input:
        "results/debarcoded/{barcode}/fastq_runid_unknown_0.fastq"
    output:
        "results/debarcoded/{barcode}/readID.txt"
    threads: 1
    resources:
        mem_mb = 2000
    run:
        for record in SeqIO.parse(input[0], "fastq"):
            with open(output[0], "a") as f:
                f.write(record.id + "\n")

rule pod5_filter:
    input:
        "results/debarcoded/{barcode}/readID.txt"
    output:
        "results/debarcoded/pod5/{barcode}.pod5"
    threads: 1
    resources:
        mem_mb = 8000
    shell:
        """
        pod5 filter -r /mnt/shared/projects/jhi/soil-ecology/202307_root2res-pilot-ONT --ids {input} --output {output}
        """
    
rule duplex:
    input:
        "results/debarcoded/pod5/{barcode}.pod5"
    output:
        "results/debarcoded/duplex/{barcode}.fastq"
    threads: 8
    params:
        dorado = config["dorado_path"],
        model = config["dorado_model"]
    resources:
        mem_mb = 128000,
        partition = "gpu",
        slurm = "gpus=2"
    shell:
        """
        {params.dorado} duplex --emite_fastq {params.model} {input} > {output}
        """